teps/0140-producing-results-in-matrix.md

```diff
@@ -0,0 +1,702 @@
+kind: PipelineRun
+metadata:
+  generateName: platforms-with-results
+spec:
+  serviceAccountName: "default"
+  pipelineSpec:
+    tasks:
+      - name: matrix-emitting-results
+        matrix:
+          params:
+            - name: platform
+              value:
+                - linux
+                - mac
+                - windows
+            - name: browser
+              value:
+                - chrome
+                - safari
+                - firefox
+        taskSpec:
+          params:
+            - name: platform
+              default: ""
+            - name: browser
+              default: ""
+          results:
+            - name: report-url
+              type: string
+          steps:
+            - name: produce-report-url
+              image: alpine
+              script: |
+                echo "Running tests on $(params.platform)-$(params.browser)"
+                echo -n "https://api.example/get-report/$(params.platform)-$(params.browser)" | tee $(results.report-url.path)
+      - name: task-consuming-results
+        taskRef:
+          name: stringtask
+          kind: Task
+        params:
+          - name: url
+            value:
+              - $(tasks.matrix-emitting-results.results.report-url[0]) # linux-chrome
+              - $(tasks.matrix-emitting-results.results.report-url[3]) # linux-safari
+              - $(tasks.matrix-emitting-results.results.report-url[6]) # linux-firefox
+
+        taskSpec:
+          params:
+            - name: url
+              type: string
+          steps:
+            - name: echo
+              image: alpine
+              script: |
+                for arg in "$@"; do
+                  echo "Arg: $arg"
+                done
+```
+
+TaskRuns Created - Result
+```
+pr-produce-platforms-and-browsers-0 - “path/to/report/linux-chrome”
+pr-produce-platforms-and-browsers-1 - “path/to/report/mac-chrome”
+pr-produce-platforms-and-browsers-2 - “path/to/report/windows-chrome”
+pr-produce-platforms-and-browsers-3 - “path/to/report/linux-safari”
+pr-produce-platforms-and-browsers-4 - “path/to/report/mac-safari”
+pr-produce-platforms-and-browsers-5 - “path/to/report/windows-safari”
+pr-produce-platforms-and-browsers-6 - “path/to/report/linux-firefox”
+pr-produce-platforms-and-browsers-7 - “path/to/report/mac-firefox”
+pr-produce-platforms-and-browsers-8 - “path/to/report/windows-firefox”
+```
+
+## Alternatives
+
+### Support only Matrix.Include Results by Result Name
+If the only use case we need to support is for explicit combinations, meaning only `Matrix.Include` without `Matrix Params`, then we can support results using the syntax of $(tasks.<include-name>.results.<result-name>) so that we can have an explicit mapping for each combination.
+Pros: This is a straightforward approach to map the combinations to the original params.
+Cons: This approach won’t support `Matrix` with `Params` or `Matrix` with both `Matrix` `Include` and `Matrix` `Params`.
```

Comment:
I like this approach. Can we expand it to support using full array results fanned out from matrix params, but not array indexing of matrixed results?